src/OVAL/oval_probe.c

```diff
@@ -350,84 +351,118 @@ int oval_probe_query_definition(oval_probe_session_t *sess, const char *id) {
-		object = oval_test_get_object(test);
-		if (object == NULL)
-			return 0;
-		/* probe object */
-		ret = oval_probe_query_object(sess, object, 0, NULL);
-		if (ret == -1)
-			return ret;
-		/* probe objects referenced like this: test->state->variable->object */
-		ste_itr = oval_test_get_states(test);
-		while (oval_state_iterator_has_more(ste_itr)) {
-			struct oval_state *state = oval_state_iterator_next(ste_itr);
-			struct oval_state_content_iterator *contents = oval_state_get_contents(state);
-			while (oval_state_content_iterator_has_more(contents)) {
-				struct oval_state_content *content = oval_state_content_iterator_next(contents);
-				struct oval_entity * entity = oval_state_content_get_entity(content);
-				if (oval_entity_get_varref_type(entity) == OVAL_ENTITY_VARREF_ATTRIBUTE) {
-					oval_syschar_collection_flag_t flag;
-					struct oval_variable *var = oval_entity_get_variable(entity);
-					const char *state_id = oval_state_get_id(state);
-					oval_variable_type_t var_type = oval_variable_get_type(var);
-					const char *var_type_text = oval_variable_type_get_text(var_type);
-					const char *var_id = oval_variable_get_id(var);
-					dI("State '%s' references %s '%s'.", state_id,
-						var_type_text, var_id);
-
-					ret = oval_probe_query_variable(sess, var);
-					if (ret == -1) {
-						oval_state_content_iterator_free(contents);
-						oval_state_iterator_free(ste_itr);
-						return ret;
-					}
-
-					flag = oval_variable_get_collection_flag(var);
-					switch (flag) {
-					case SYSCHAR_FLAG_COMPLETE:
-					case SYSCHAR_FLAG_INCOMPLETE:
-						break;
-					default:
-						oval_state_content_iterator_free(contents);
-						oval_state_iterator_free(ste_itr);
-						return 0;
-					}
+	rslt_test = oval_result_criteria_node_get_test(rslt_cnode);
+	test = oval_criteria_node_get_test(cnode);
+	if (test == NULL) {
+		return oval_result_test_eval(rslt_test);
+	}
+	type = oval_subtype_get_text(oval_test_get_subtype(test));
+	test_id = oval_test_get_id(test);
+	comment = oval_test_get_comment(test);
+	dI("Evaluating %s test '%s': %s.", type, test_id, comment);
+	object = oval_test_get_object(test);
+	if (object == NULL) {
+		return oval_result_test_eval(rslt_test);
+	}
+
+	/* probe object */
+	ret = oval_probe_query_object(sess, object, 0, NULL);
+	if (ret == -1) {
+		return oval_result_test_eval(rslt_test);
+	}
+
+	/* probe objects referenced like this: test->state->variable->object */
+	ste_itr = oval_test_get_states(test);
+	while (oval_state_iterator_has_more(ste_itr)) {
+		struct oval_state *state = oval_state_iterator_next(ste_itr);
+		struct oval_state_content_iterator *contents = oval_state_get_contents(state);
+		while (oval_state_content_iterator_has_more(contents)) {
+			struct oval_state_content *content = oval_state_content_iterator_next(contents);
+			struct oval_entity * entity = oval_state_content_get_entity(content);
+			if (oval_entity_get_varref_type(entity) == OVAL_ENTITY_VARREF_ATTRIBUTE) {
```

Comment:
I would consider to move IF block to extra function. Very hard to read, many variables, long function.